Port to GTK3 and Python3

Saumya-Mishra9129 opened the pull request:
Tested , works as expected. draw by clicking on stop button is solved in bf4201f. I did what @quozl suggested in https://github.com/sugarlabs/starchart/pull/6#issuecomment-406909794. @quozl @chimosky Please review.

yell0wfl4sh committed:
port starchart.py

yell0wfl4sh committed:
update setup.py

yell0wfl4sh committed:
make window work

yell0wfl4sh committed:
add pep8 fixes

yell0wfl4sh committed:
add idea in .gitignore

yell0wfl4sh committed:
add fills and strokes

yell0wfl4sh committed:
update arc

yell0wfl4sh committed:
fill color

yell0wfl4sh committed:
fix starsize and many more

yell0wfl4sh committed:
remove comments

yell0wfl4sh committed:
change to ToggleButtonTool

yell0wfl4sh committed:
remove elipse code

yell0wfl4sh committed:
fix logged errors

yell0wfl4sh committed:
fix cross width and color

yell0wfl4sh committed:
make grey objects grey again

Saumya-Mishra9129 committed:
Merge branch 'gtk3' of https://github.com/yashagrawal3/starchart

Saumya-Mishra9129 committed:
Run 2to3

Saumya-Mishra9129 committed:
Remove Unnecessary backslash in expression

Saumya-Mishra9129 committed:
change exec to sugar-activity3

Saumya-Mishra9129 committed:
change shebang to python3

Saumya-Mishra9129 committed:
fix draw event called at the time of pressing stop

Use Gdk.DrawingArea via context (c) -- Rename area_expose_cb to draw_cb , fix argument names for draw_cb

quozl commented:
Thanks.  Reviewed by examining diff, with and without whitespace changes.

* [ ] the test for the old toolbox is no longer required,
* [ ] documentation pointer to wiki.laptop.org should be replaced,

Tested for excessive updates;

```
GTK_DEBUG=updates sugar-activity3
```

Pass.  bf4201f is effective.

Tested features;

* [ ] all the view changing buttons are initially ineffective; they should queue a redraw,

Checked appearance;

* [ ] line drawing width regression; the constellation lines and stars are different size; they should be the same size as the Python 2 version shows, see screenshots;

20b85cea354af351d72873a923196759e7512261
![Screenshot_ubb_2020-07-21_09:42:31](https://user-images.githubusercontent.com/2936481/87996813-a7ac9780-cb36-11ea-95a5-15b29ced9ebe.png)

bf4201fb3fb073b172f33bc5e38749ac912bed48
![Screenshot_uff_2020-07-21_09:42:40](https://user-images.githubusercontent.com/2936481/87996826-af6c3c00-cb36-11ea-9140-0ede9c381b26.png)

Saumya-Mishra9129 commented:
> all the view changing buttons are initially ineffective; they should queue a redraw,

I saw some ambiguous behaviour of view buttons. They are queuing redraw, but only when we click on main view button after clicking on a particular view button.

> line drawing width regression; the constellation lines and stars are different size; they should be the same size as the Python 2 version shows, see screenshots;

Can you please tell on which display you are testing?

Saumya-Mishra9129 committed:
Remove support for old toolbox

Saumya-Mishra9129 committed:
Fix documentation link in Gtk.Label

quozl commented:
> > all the view changing buttons are initially ineffective; they should queue a redraw,
> 
> I saw some ambiguous behaviour of view buttons. They are queuing redraw, but only when we click on main view button after clicking on a particular view button.

Why does that happen?

> > line drawing width regression; the constellation lines and stars are different size; they should be the same size as the Python 2 version shows, see screenshots;
> 
> Can you please tell on which display you are testing?

In the above screenshots, QEMU/KVM with virtual display configured for 1024x768 resolution.  I don't understand why you need to know this.

Saumya-Mishra9129 commented:
> Why does that happen?

I don't know. I haven't made any change which could result in this behaviour. You have also mentioned the issue in https://github.com/sugarlabs/starchart/pull/6#issuecomment-402355771. 

`buttons night vision, invert display, flip l/r, draw constellations, the magnitude radio buttons, latitude changes, all time changes; do not instantly redraw; workaround is to collapse the toolbar,`.

> In the above screenshots, QEMU/KVM with virtual display configured for 1024x768 resolution. I don't understand why you need to know this.

Actually this issue is also introduced by that previous port to Gtk3 , Trying to look into it what could cause this behaviour.

Saumya-Mishra9129 committed:
Remove some global variables : use with-in class variables

quozl commented:
> > Why does that happen?
> 
> I don't know. I haven't made any change which could result in this behaviour. You have also mentioned the issue in [#6 (comment)](https://github.com/sugarlabs/starchart/pull/6#issuecomment-402355771).

It is caused by not calling `Gtk.Widget.queue_draw()` in the _clicked_ callback of the night vision button.  In turn, caused by not properly porting from GTK 2.  `Gtk.DrawingArea` is very different, and drawing methods very different.  The draw callback in your branch doesn't properly use the supplied Cairo context argument; it is saved for later use by the button callbacks, by which time the (copy of) the context is unlikely to be valid.  Drawing is most properly done within the draw callback, and if any button callbacks need to activate the draw callback they must call _queue_draw_.

Summary is that the Port to GTK 3 is incomplete.

JuiP commented:
Tested not yet reviewed changes; some additional observations/suggestions:

- [ ] The activity doesnot implement collaboration so set `self.max_participants = 1`

- [ ] Consider setting tooltips for activity toolbar icons

- [ ] All the view buttons were ineffective at start. All the changes seem to be applied each time I click on the longitude button.

- [ ] Select Object Type, Brightest Stars and choose one option eg. Adharaz , a green cross appears. I am assuming it shows the position of the selected star. Now use flip L/R option under view; the image is flipped but the green cross still points to the same area, effectively now it points to the wrong star. See below screenshots:

![con-1](https://user-images.githubusercontent.com/46154968/89715246-0d67a300-d9c2-11ea-9ce4-76b3012117cb.png)

After flip points to wrong star:
![flipped](https://user-images.githubusercontent.com/46154968/89715250-0fc9fd00-d9c2-11ea-879c-a0bcfa4926c8.png)

chimosky commented:
@quozl said;
> It is caused by not calling `Gtk.Widget.queue_draw()` in the _clicked_ callback of the night vision button. In turn, caused by not properly porting from GTK 2. `Gtk.DrawingArea` is very different, and drawing methods very different. The draw callback in your branch doesn't properly use the supplied Cairo context argument; it is saved for later use by the button callbacks, by which time the (copy of) the context is unlikely to be valid. Drawing is most properly done within the draw callback, and if any button callbacks need to activate the draw callback they must call _queue_draw_.
> 
> Summary is that the Port to GTK 3 is incomplete.

Fixed `queue_draw` not being called in draw_cb.

@JuiP said;
>  - [x]  The activity doesnot implement collaboration so set self.max_participants = 1
>  - [ ]  Consider setting tooltips for activity toolbar icons

When tooltips are set, clicking on the buttons doesn't activate them when they're clicked for some reason.

>  - [ ] Select Object Type, Brightest Stars and choose one option eg. Adharaz , a green cross appears. I am assuming it shows the position of the selected star. Now use flip L/R option under view; the image is flipped but the green cross still points to the same area, effectively now it points to the wrong star. 

This is an existing bug and wasn't created by this PR, doesn't mean it can't be fixed here but I don't think it's paramount that it's fixed here.

quozl commented:
f5f7b7b ("queue draw in draw_cb") causes an infinite loop because every time GTK asks the app to draw, the app asks GTK to ask to draw again.

It is not correct to call `queue_draw` inside the draw callback.  Instead, `queue_draw` must be called in each place that requires a redraw, such as the _clicked_ callback of the night vision button.

chimosky committed:
Set max_participants to 1

Signed-off-by: Ibiam Chihurumnaya <ibiamchihurumnaya@gmail.com>

chimosky commented:
> [f5f7b7b](https://github.com/sugarlabs/starchart/commit/f5f7b7b4725df183fe3ec18f756a5f7bb094a1f0) ("queue draw in draw_cb") causes an infinite loop because every time GTK asks the app to draw, the app asks GTK to ask to draw again.
> 
> It is not correct to call `queue_draw` inside the draw callback. Instead, `queue_draw` must be called in each place that requires a redraw, such as the _clicked_ callback of the night vision button.

Seems it's been called in all the necessary places, do you still notice the appearance difference in your tests?

quozl commented:
Tetsed f4933fe.  The view buttons remain ineffective.  Here's how I fixed it for the Flip L/R button;

```diff
diff --git a/StarChart.py b/StarChart.py
index 225ec62..f643714 100644
--- a/StarChart.py
+++ b/StarChart.py
@@ -1164,6 +1164,7 @@ class ChartDisplay(Gtk.DrawingArea):
         elif (data == 'flip horizontally'):
             self.fliphorizontally = button3.get_active()
             self.plotchart()
+            self.queue_draw()
             return False
         elif (data == 'draw constellations'):
             self.drawconstellations = button4.get_active()
```

Probably there is a better place to put this.

chimosky committed:
Remove redundant try-except clause

switch to floor division when divisor is an int
remove redundant reversed call in loops
call self.queue_draw in plotchart for a redraw

Signed-off-by: Ibiam Chihurumnaya <ibiamchihurumnaya@gmail.com>
